File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/Opaque.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Opaque {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class OpaqueCleaner implements Runnable {

        MemorySegment segment;
        OpaqueCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Opaque_destroy(this.segment);
        }
    }

    Opaque() {}
    Opaque(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    public static Opaque new_() {
        
        var nativeVal = somelib_h.Opaque_new();
        
        List<Object> selfEdges = List.of();
        
        
        
        var returnVal = new Opaque(nativeVal, selfEdges);
        return returnVal;
                
    }
    
    public static Opaque fromStr(String input) {
        
        try (var arena = Arena.ofConfined()) {
            var inputData= Arena.ofAuto().allocateFrom(input, StandardCharsets.UTF_8);
            var inputLen = inputData.byteSize() - 1;  // allocated strings are null terminated
            var inputView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(inputView, inputLen);
            DiplomatStringView.data(inputView, inputData);
            var nativeVal = somelib_h.Opaque_from_str(inputView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Opaque(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    public static long returnsUsize() {
        
        var nativeVal = somelib_h.Opaque_returns_usize();
        
        var returnVal = nativeVal;
        return returnVal;
                
    }
    
    public static ImportedStruct returnsImported() {
        
        try (var arena = Arena.ofConfined()) {
            
            var nativeVal = somelib_h.Opaque_returns_imported(arena);
            
            var returnVal = new ImportedStruct(nativeVal);
            return returnVal;
                    
        }
    }
    
    public static byte cmp() {
        
        var nativeVal = somelib_h.Opaque_cmp();
        
        var returnVal = nativeVal;
        return returnVal;
                
    }
    
    
    public String getDebugStr() {
        
        
        var writeable = somelib_h.diplomat_buffer_write_create(0);
        somelib_h.Opaque_get_debug_str(internal, writeable);
        var string = SliceUtils. readUtf8FromWriteable(writeable);
        somelib_h.diplomat_buffer_write_destroy(writeable);
        return string;
    }
    
    public void assertStruct(MyStruct s) {
        
        
        var sNative = s.toNative(Arena.ofAuto());
        somelib_h.Opaque_assert_struct(internal, sNative);
        
    }
    
    public long internalLen() {
        
        
        var nativeVal = somelib_h.Opaque_internal_len(internal);
        
        var returnVal = nativeVal;
        return returnVal;
                
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/Opaque.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface OpaqueLib: Library {
    fun Opaque_destroy(handle: Pointer)
    fun Opaque_new(): Pointer
    fun Opaque_try_from_utf8(input: Slice): Pointer?
    fun Opaque_from_str(input: Slice): Pointer
    fun Opaque_get_debug_str(handle: Pointer, write: Pointer): Unit
    fun Opaque_assert_struct(handle: Pointer, s: MyStructNative): Unit
    fun Opaque_returns_usize(): FFISizet
    fun Opaque_returns_imported(): ImportedStructNative
    fun Opaque_cmp(): Byte
}

class Opaque internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
)  {

    internal class OpaqueCleaner(val handle: Pointer, val lib: OpaqueLib) : Runnable {
        override fun run() {
            lib.Opaque_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<OpaqueLib> = OpaqueLib::class.java
        internal val lib: OpaqueLib = Native.load("somelib", libClass)
        @JvmStatic
        
        fun new_(): Opaque {
            
            val returnVal = lib.Opaque_new();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Opaque(handle, selfEdges)
            CLEANER.register(returnOpaque, Opaque.OpaqueCleaner(handle, Opaque.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun tryFromUtf8(input: String): Opaque? {
            val (inputMem, inputSlice) = PrimitiveArrayTools.borrowUtf8(input)
            
            val returnVal = lib.Opaque_try_from_utf8(inputSlice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal ?: return null
            val returnOpaque = Opaque(handle, selfEdges)
            CLEANER.register(returnOpaque, Opaque.OpaqueCleaner(handle, Opaque.lib));
            if (inputMem != null) inputMem.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun fromStr(input: String): Opaque {
            val (inputMem, inputSlice) = PrimitiveArrayTools.borrowUtf8(input)
            
            val returnVal = lib.Opaque_from_str(inputSlice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Opaque(handle, selfEdges)
            CLEANER.register(returnOpaque, Opaque.OpaqueCleaner(handle, Opaque.lib));
            if (inputMem != null) inputMem.close()
            return returnOpaque
        }
        @JvmStatic
        
        fun returnsUsize(): ULong {
            
            val returnVal = lib.Opaque_returns_usize();
            return (returnVal.toULong())
        }
        @JvmStatic
        
        fun returnsImported(): ImportedStruct {
            
            val returnVal = lib.Opaque_returns_imported();
            
            val returnStruct = ImportedStruct(returnVal)
            return returnStruct
        }
        @JvmStatic
        
        fun cmp(): Byte {
            
            val returnVal = lib.Opaque_cmp();
            return (returnVal)
        }
    }
    
    fun getDebugStr(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Opaque_get_debug_str(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    /** See the [Rust documentation for `something`](https://docs.rs/Something/latest/struct.Something.html#method.something) for more information.
    *
    *See the [Rust documentation for `something_else`](https://docs.rs/Something/latest/struct.Something.html#method.something_else) for more information.
    *
    *Additional information: [1](https://docs.rs/Something/latest/struct.Something.html#method.something_small), [2](https://docs.rs/SomethingElse/latest/struct.SomethingElse.html#method.something)
    */
    fun assertStruct(s: MyStruct): Unit {
        
        val returnVal = lib.Opaque_assert_struct(handle, s.nativeStruct);
        
    }

}